File: processor/src/main/java/com/qualityminds/lazyval/processor/internal/PackageLookup.java
package com.qualityminds.lazyval.processor.internal;

import org.jspecify.annotations.Nullable;

import java.util.function.Supplier;

/**
 * Resolves the package a generator writes to, preferring its own override over the global
 * base-package default and falling back to a supplier when neither is configured.
 *
 * @see com.qualityminds.lazyval.processor.spi.Generator.Context#generatorPackage
 */
class PackageLookup {

    record DefaultConfig(String basePackage, @Nullable String layer){
        static @Nullable DefaultConfig of(@Nullable String basePackage, @Nullable String layer){
            if(basePackage == null){
                return null;
            }
            return new DefaultConfig(basePackage, layer);
        }

        String defaultPackage() {
            if(layer != null){
                return basePackage + "." + layer;
            }
            return basePackage;
        }
    }

    static String computePackage(@Nullable DefaultConfig config, @Nullable String generatorOverridePackage, Supplier<String> fallbackSupplier){
        if(generatorOverridePackage != null){
            return generatorOverridePackage;
        }else if(config != null){
            return config.defaultPackage();
        } else {
            return fallbackSupplier.get();
        }
    }

    private PackageLookup(){}
}

File: ksp/src/main/kotlin/com/qualityminds/lazyval/ksp/internal/PackageLookup.kt
package com.qualityminds.lazyval.ksp.internal

import java.util.function.Supplier

/**
 * @see com.qualityminds.lazyval.ksp.spi.Generator.Context.generatorPackage
 */
object PackageLookup {

    @JvmRecord
    data class DefaultConfig(val basePackage: String, val layer: String?) {
        fun defaultPackage(): String {
            if (layer != null) {
                return "$basePackage.$layer"
            }
            return basePackage
        }

        companion object {
            @JvmStatic
            fun of(basePackage: String?, layer: String?): DefaultConfig? {
                if (basePackage == null) {
                    return null
                }
                return DefaultConfig(basePackage, layer)
            }
        }
    }

    @JvmStatic
    fun computePackage(
        config: DefaultConfig?,
        generatorOverridePackage: String?,
        fallbackSupplier: Supplier<String>
    ): String {
        return generatorOverridePackage
            ?: (config?.defaultPackage() ?: fallbackSupplier.get())
    }

}